src/ioctls/common/kvm_vec.rs

```diff
@@ -0,0 +1,549 @@
+            Vec::from_raw_parts(entries_ptr, self.len, self.capacity as usize)
+        }
+    }
+
+    /// Get the mutable elements slice so they can be modified before passing to the VCPU.
+    ///
+    pub fn as_entries_slice(&self) -> &[T::Entry] {
+        let len = self.as_kvm_struct().len();
+        unsafe { self.as_kvm_struct().entries().as_slice(len as usize) }
+    }
+
+    /// Get the mutable elements slice so they can be modified before passing to the VCPU.
+    ///
+    pub fn as_mut_entries_slice(&mut self) -> &mut [T::Entry] {
+        let len = self.as_kvm_struct().len();
+        unsafe {
+            self.as_mut_kvm_struct()
+                .entries_mut()
+                .as_mut_slice(len as usize)
+        }
+    }
+
+    /// Reserves capacity for at least `additional` more `KvmArray::Entry` elements.
+    /// If the capacity is already reserved, this method doesn't do anything
+    ///
+    fn reserve(&mut self, additional: usize) {
+        let desired_capacity = self.len + additional;
+        if desired_capacity <= self.capacity {
+            return;
+        }
+
+        let current_mem_allocator_len = self.mem_allocator.len();
+        let required_mem_allocator_len =
+            KvmVec::<T>::kvm_vec_len_to_mem_allocator_len(desired_capacity);
+        let additional_mem_allocator_len = required_mem_allocator_len - current_mem_allocator_len;
+
+        self.mem_allocator.reserve(additional_mem_allocator_len);
+        self.capacity =
+            KvmVec::<T>::mem_allocator_len_to_kvm_vec_len(self.mem_allocator.capacity());
+    }
+
+    /// Updates the length of `self` to the specified value.
+    /// Also updates the length of the `T::Entry` structure and of `self.mem_allocator` accordingly.
+    ///
+    fn update_len(&mut self, len: usize) {
+        self.len = len;
+        self.as_mut_kvm_struct().set_len(len);
+
+        /// We need to set the len of the mem_allocator to be the number of T elements needed
+        /// to fit an array of `len` elements of type `T::Entry`. This way, when we call
+        /// `self.mem_allocator.shrink_to_fit()` only the unnecessary memory will be released.
+        let required_mem_allocator_len = KvmVec::<T>::kvm_vec_len_to_mem_allocator_len(len);
+        unsafe {
+            self.mem_allocator.set_len(required_mem_allocator_len);
+        }
+    }
+
+    /// Appends an element to the end of the collection and updates `len`.
+    ///
+    /// # Arguments
+    ///
+    /// * `entry` - The element that will be appended to the end of the collection.
+    ///
+    /// # Error: When len is already equal to max possible len it returns Error::SizeLimitExceeded
+    ///
+    pub fn push(&mut self, entry: T::Entry) -> Result<(), Error> {
+        let desired_len = self.len + 1;
+        if desired_len > T::max_len() {
+            return Err(Error::SizeLimitExceeded);
+        }
+
+        self.reserve(1);
+
+        let mut entries = self.as_vec();
```

Comment:
Is the `self.as_vec`/`push` combination required here? Since we've already ensured there is enough space via `reserve`, it looks like we can call `self.update_len(desired_len);`, and then update the value of the last position based on `entry` using the mut slice returned by `self.as_mut_entries_slice`.